# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [81]:
#Import your libraries
import pandas as pd
import numpy as np
from datetime import datetime as dt



# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. You can find the data in Ironhack's database (db: weather, table: austin_weather). 

In [33]:
# Your code here
austin = pd.read_csv('austin_weather.csv', sep= ',')
austin.head(10)


Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   
5  2013-12-26         57        48        39            39           36   
6  2013-12-27         60        53        45            41           39   
7  2013-12-28         62        51        40            43           39   
8  2013-12-29         64        50        36            49           41   
9  2013-12-30         44        40        35            31           26   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   
5           33                  79                 63                 47  ...   
6           37                  83                 65                 47  ...   
7           33                  92                 64                 36  ...   
8           28                  92                 76                 60  ...   
9           21                  75                 60                 45  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   
5                      30.4                     30.34                  10   
6                     30.39                     30.34                  10   
7                     30.17                     30.04                  10   
8                      30.1                     29.99                  10   
9                     30.33                     30.26                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   
5                  9                  7          12          3          17   
6                  9                  7           7          1          11   
7                 10                  7          10          2          14   
8                 10                  4          17          5          24   
9                 10                 10          13          5          21   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       
5                      0                       
6                      T                       
7                      T                       
8                      0                       
9                      0                       

[10 rows x 21 columns]

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [34]:
# variables and their types
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1319 non-null object
DewPointAvgF                  1319 non-null object
DewPointLowF                  1319 non-null object
HumidityHighPercent           1319 non-null object
HumidityAvgPercent            1319 non-null object
HumidityLowPercent            1319 non-null object
SeaLevelPressureHighInches    1319 non-null object
SeaLevelPressureAvgInches     1319 non-null object
SeaLevelPressureLowInches     1319 non-null object
VisibilityHighMiles           1319 non-null object
VisibilityAvgMiles            1319 non-null object
VisibilityLowMiles            1319 non-null object
WindHighMPH                   1319 non-null object
WindAvgMPH                   

In [35]:
# descriptive statistics of the numeric variables
austin.describe()


TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [36]:
#irst five rows of all variables to evaluate the categorical variables 
austin.head()


Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [37]:
# Your code here
austin['DewPointAvgF'].unique()


array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)


type of variables set as object

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [38]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [39]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce', axis=1)

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [40]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1312 non-null float64
DewPointAvgF                  1312 non-null float64
DewPointLowF                  1312 non-null float64
HumidityHighPercent           1317 non-null float64
HumidityAvgPercent            1317 non-null float64
HumidityLowPercent            1317 non-null float64
SeaLevelPressureHighInches    1316 non-null float64
SeaLevelPressureAvgInches     1316 non-null float64
SeaLevelPressureLowInches     1316 non-null float64
VisibilityHighMiles           1307 non-null float64
VisibilityAvgMiles            1307 non-null float64
VisibilityLowMiles            1307 non-null float64
WindHighMPH                   1317 non-null float64
WindAvgMPH      

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [41]:
# Your code here
missing_values = austin[austin.isnull().any(axis=1)]
missing_values

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4     2013-12-25         58        50        41           44.0          40.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
42    2014-02-01         76        66        55           62.0          59.0   
51    2014-02-10         60        48        35           49.0          36.0   
66    2014-02-25         71        62        52           65.0          60.0   
95    2014-03-26         66        60        54           55.0          47.0   
102   2014-04-02         82        76        70           69.0          67.0   
103   2014-04-03         82        77        71           69.0          67.0   
104   2014-04-04         74        64        54           69.0          35.0   
105   2014-04-05         65        60        54           46.0          38.0   
113   2014-04-13         78        73        68           68.0          66.0   
126   2014-04-26         86        77        68           69.0          66.0   
139   2014-05-09         86        76        66           71.0          68.0   
152   2014-05-22         87        79        70           68.0          65.0   
154   2014-05-24         86        78        70           70.0          68.0   
155   2014-05-25         87        80        72           67.0          66.0   
164   2014-06-03         90        83        75           72.0          70.0   
169   2014-06-08         90        83        76           72.0          70.0   
174   2014-06-13         89        79        68            NaN           NaN   
175   2014-06-14         87        92        73            NaN           NaN   
176   2014-06-15         91        83        74            NaN           NaN   
177   2014-06-16         92        84        75            NaN           NaN   
182   2014-06-21         92        84        75           74.0          72.0   
183   2014-06-22         90        83        76           74.0          72.0   
188   2014-06-27         91        83        75           74.0          72.0   
189   2014-06-28         93        86        78           74.0          73.0   
248   2014-08-26        101        89        77           71.0          68.0   
278   2014-09-25         83        73        62           66.0          62.0   
280   2014-09-27         80        74        68           69.0          66.0   
...          ...        ...       ...       ...            ...           ...   
1101  2016-12-26         84        77        70           70.0          69.0   
1102  2016-12-27         76        69        62           68.0          58.0   
1105  2016-12-30         55        52        49           34.0          23.0   
1106  2016-12-31         77        64        51           61.0          51.0   
1107  2017-01-01         73        63        53           60.0          52.0   
1125  2017-01-19         71        62        52           55.0          52.0   
1134  2017-01-28         61        50        39           37.0          31.0   
1140  2017-02-03         53        49        45           37.0          34.0   
1142  2017-02-05         72        62        51           64.0          59.0   
1143  2017-02-06         78        73        67           67.0          65.0   
1150  2017-02-13         78        71        63           64.0          55.0   
1163  2017-02-26         72        57        42           64.0          50.0   
1164  2017-02-27         82        74        66           66.0          65.0   
1171  2017-03-06         84        76        68           67.0          66.0   
1172  2017-03-07         77        69        61           67.0          51.0   
1173  2017-03-08         76        68        60           64.0          50.0   
1174  2017-03-09         80        74        67           66.0          64.0   
1175  2017-03-10         83        73        62           65.0          62.0   
1209  2017-0

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [42]:
# Your code here
print(len(austin))
print(len(missing_values))


1319
136


#### Calculate the ratio of missing rows to total rows

In [43]:
# Your code here
len(austin)/len(missing_values)

9.698529411764707

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [44]:
# Your code here
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [45]:
# ratio of missing values in 'PrecipitationSumInches' to total number

austin['PrecipitationSumInches'].isna().sum()/len(austin)


0.09401061410159212

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [46]:
# Your code here 

austin.drop(['PrecipitationSumInches'], axis=1, inplace=True) 
# Print `austin` to confirm the column is indeed removed
austin.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                Events  
0  Rain , Thunderstorm  
1                       
2                       
3                       
4

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [16]:
# Your code here
austin_fixed = austin.interpolate(inplace=False)


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [47]:
# Your code here
print(austin_fixed.info())
print(austin.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 24 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1319 non-null float64
DewPointAvgF                  1319 non-null float64
DewPointLowF                  1319 non-null float64
HumidityHighPercent           1319 non-null float64
HumidityAvgPercent            1319 non-null float64
HumidityLowPercent            1319 non-null float64
SeaLevelPressureHighInches    1319 non-null float64
SeaLevelPressureAvgInches     1319 non-null float64
SeaLevelPressureLowInches     1319 non-null float64
VisibilityHighMiles           1319 non-null float64
VisibilityAvgMiles            1319 non-null float64
VisibilityLowMiles            1319 non-null float64
WindHighMPH                   1319 non-null float64
WindAvgMPH      

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [48]:
# Your code here:
austin_fixed['Events'].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [89]:
# Your answer:
'3'

'3'

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [49]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
for event in event_list:
    austin_fixed[event] = 0
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     0            0  
1         25.0                          0    0     0            0  
2         12.0                          0    0     0            0  
3         20.0                          0    0     0            0  
4         16.0                          0    0     0            0  

[5 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [50]:
# Your code here
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
for event in event_list:
    filt = austin_fixed['Events'].str.contains(event)
    austin_fixed.loc[filt, event] = 1




#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [88]:
# Your code here
austin_fixed.head(100)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0   719163         74        60        45           67.0          49.0   
1   719163         56        48        39           43.0          36.0   
2   719163         58        45        32           31.0          27.0   
3   719163         61        46        31           36.0          28.0   
4   719163         58        50        41           44.0          40.0   
5   719163         57        48        39           39.0          36.0   
6   719163         60        53        45           41.0          39.0   
7   719163         62        51        40           43.0          39.0   
8   719163         64        50        36           49.0          41.0   
9   719163         44        40        35           31.0          26.0   
10  719163         55        46        36           31.0          28.0   
11  719163         69        54        39           51.0          42.0   
12  719163         55        44        33           39.0          26.0   
13  719163         58        43        28           37.0          22.0   
14  719163         71        57        42           55.0          48.0   
15  719163         59        47        34           54.0          32.0   
16  719163         36        29        22           15.0           8.0   
17  719163         48        35        22           29.0          11.0   
18  719163         53        47        40           51.0          45.0   
19  719163         70        62        53           60.0          55.0   
20  719163         72        65        57           64.0          61.0   
21  719163         75        62        48           54.0          37.0   
22  719163         67        57        46           58.0          47.0   
23  719163         71        57        43           60.0          36.0   
24  719163         76        57        37           31.0          25.0   
25  719163         64        52        40           29.0          23.0   
26  719163         72        56        40           31.0          27.0   
27  719163         64        54        43           31.0          20.0   
28  719163         70        54        37           41.0          33.0   
29  719163         72        59        45           41.0          30.0   
..     ...        ...       ...       ...            ...           ...   
70  719163         82        68        53           65.0          57.0   
71  719163         73        51        28           67.0          45.0   
72  719163         47        36        24           27.0          20.0   
73  719163         50        40        30           36.0          31.0   
74  719163         53        46        38           43.0          39.0   
75  719163         65        54        43           42.0          38.0   
76  719163         67        53        39           50.0          42.0   
77  719163         59        52        44           57.0          49.0   
78  719163         58        49        39           45.0          39.0   
79  719163         71        61        50           56.0          51.0   
80  719163         83        70        56           56.0          54.0   
81  719163         70        60        49           50.0          28.0   
82  719163         68        52        37           40.0          26.0   
83  719163         75        62        49           59.0          51.0   
84  719163         74        68        62           64.0          62.0   
85  719163         64        56        48           58.0          45.0   
86  719163         68        55        41           31.0          26.0   
87  719163         83        65        47           56.0          42.0   
88  719163         72        62        51           55.0          37.0   
89  719163         76        59        42           40.0          36.0   
90  719163         80        67        54           60.0          54.0   
91  719163         77        69        61           65.0          62.0   
92  719163

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [52]:
# Your code here

austin.drop(['Events'], axis=1, inplace=True) 

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [86]:
# Your code here

austin_fixed['Date'] = pd.to_datetime(austin_fixed['Date'])
austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: x.toordinal())

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  719163         74        60        45           67.0          49.0   
1  719163         56        48        39           43.0          36.0   
2  719163         58        45        32           31.0          27.0   
3  719163         61        46        31           36.0          28.0   
4  719163         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     1            1  
1         25.0                          0    0     0            0  
2         12.0                          0    0     0            0  
3         20.0                          0    0     0            0  
4         16.0                          0    0     0            0  

[5 rows x 24 columns]

#### Print `austin_fixed` to check your `Date` column.

In [87]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  719163         74        60        45           67.0          49.0   
1  719163         56        48        39           43.0          36.0   
2  719163         58        45        32           31.0          27.0   
3  719163         61        46        31           36.0          28.0   
4  719163         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     1            1  
1         25.0                          0    0     0            0  
2         12.0                          0    0     0            0  
3         20.0                          0    0     0            0  
4         16.0                          0    0     0            0  

[5 rows x 24 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [90]:
# Your code here:
y = austin_fixed[['TempAvgF']]
X = austin_fixed[['TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]


In the next cell, import `train_test_split` from `sklearn.model_selection`

In [64]:
#Your code here:
from sklearn.model_selection import train_test_split 

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [93]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8,test_size=0.2)
print ("X_train: ", X_train)


X_train:        TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
522          85        67           75.0          70.0          64.0   
1142         72        51           64.0          59.0          51.0   
1307        103        77           75.0          69.0          61.0   
843          73        62           70.0          61.0          58.0   
957          99        76           75.0          72.0          70.0   
921          96        71           75.0          70.0          66.0   
76           67        39           50.0          42.0          36.0   
45           64        42           45.0          41.0          39.0   
1138         82        51           59.0          52.0          45.0   
294          78        59           72.0          62.0          55.0   
130          75        57           38.0          33.0          31.0   
7            62        40           43.0          39.0          33.0   
88           72        51           55.0          37.0

In [94]:
print ("y_train: ", y_train)


y_train:        TempAvgF
522         76
1142        62
1307        90
843         68
957         88
921         84
76          53
45          53
1138        67
294         69
130         66
7           51
88          62
1030        85
84          68
909         87
66          62
1000        85
1271        87
189         86
828         62
794         54
1318        88
464         72
65          66
1273        88
804         61
1024        68
367         51
1054        63
...        ...
643         84
446         63
1301        89
831         76
914         86
932         87
692         61
171         78
981         83
1140        49
308         75
342         57
1155        74
74          46
1035        65
952         87
612         89
235         83
778         50
845         66
983         83
145         65
1113        32
1166        64
1020        84
977         85
1296        89
934         89
857         79
1171        76

[1055 rows x 1 columns]


In [95]:
print("X_test: ", X_test)


X_test:        TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
788          75        50           42.0         33.00          25.0   
233         102        76           75.0         71.00          66.0   
1055         71        56           58.0         55.00          53.0   
1010         73        67           72.0         69.00          62.0   
176          91        74           72.8         70.20          66.4   
834          79        44           45.0         41.00          39.0   
201          96        72           71.0         68.00          63.0   
190          95        77           74.0         73.00          70.0   
937         101        78           75.0         71.00          64.0   
819          62        47           49.0         36.00          28.0   
561          91        76           75.0         73.00          71.0   
1154         77        50           60.0         50.00          41.0   
6            60        45           41.0         39.00 

In [96]:
print ("y_test: ", y_test)

y_test:        TempAvgF
788         63
233         89
1055        64
1010        70
176         83
834         62
201         84
190         86
937         90
819         55
561         84
1154        64
6           53
1199        70
466         76
743         52
1100        74
1261        80
678         71
782         68
1083        44
386         42
1238        76
414         68
104         64
581         87
726         54
966         87
1107        63
1101        77
...        ...
105         60
1185        75
1027        79
596         89
213         86
1285        82
577         86
274         81
79          61
451         70
550         83
151         80
687         62
766         51
279         76
606         89
772         66
291         82
1275        89
742         45
745         46
244         91
793         65
632         77
865         71
959         89
1181        69
943         89
403         62
965         92

[264 rows x 1 columns]


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [29]:
# Your code here:
ts_rows = 


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [30]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [31]:
# Your code here:
